streams/src/main/java/org/apache/kafka/streams/query/QueryResult.java

```diff
@@ -197,6 +197,18 @@ public R getResult() {
         return result;
     }
 
+    @SuppressWarnings("unchecked")
+    public <V> QueryResult<V> swapResult(final V value) {
+        if (isFailure()) {
+            return (QueryResult<V>) this;
```

Comment:
Why do we know allow to swap the result if the current result has a failure? And if we don't want to allow swapping, why just return `this` but not throw an exception?

Reply:
In the case of a failure, there is no result, just the failure message. I wanted to maintain an invariant that there is always either a failure or a result, but not both or neither. I also didn't think it would be right to allow accidentally converting a failure to a successful result via this method.

I suppose we could throw an IllegalStateException, since the caller probably shouldn't be even trying to "swap" the result on a failed result to begin with, but this seems fine, too.

Reply:
> I suppose we could throw an IllegalStateException, since the caller probably shouldn't be even trying to "swap" the result on a failed result to begin with, but this seems fine, too.

If there is no strict reason not to throw an `IllegalStateException`, I would strongly advocate to throw. It's not guards against potential bugs, but also expressed the semantics for developers (ie, us) much cleaner and make the code easier to read/reason about.

Reply:
sounds good!